### PR TITLE
Add ability to parse just until the first directive

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue - #1195 

@rynowak @NTaylorMullen 

Super hacky design PR. Some variable cleanup got mixed up in the middle of this. Please ignore those.

### CODE REVIEW

**Comment by ajaybhargavb** [comment]
🆙 📅 Using options to get the boolean. 
@rynowak I didn't do what you suggested (Putting the DirectiveTokenizer logic in CSharpTokenizer itself) because it turned out to be a fair amount of "if"s and didn't look very good. Let me know if you feel strongly otherwise.

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/RazorParser.cs [hunk 1]
```diff
@@ -37,7 +37,10 @@ public virtual RazorSyntaxTree Parse(RazorSourceDocument source)
 
             var reader = new SeekableTextReader(chars, source.FileName);
 
-            var context = new ParserContext(reader, Options.DesignTimeMode);
+            var context = new ParserContext(reader, Options.DesignTimeMode)
+            {
+                StopParsingAfterFirstDirective = Options.StopParsingAfterFirstDirective
+            };
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/RazorParser.cs]: This should be a separate method, don't expose this on options

Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/RazorParser.cs]: Hmm, we wouldn't add a parse single directive method to RazorSyntaxTree correct? Just trying to validate that we don't add noise to currently super simple APIs.

Currently the options approach is a somewhat hidden approach that MVC can utilize and not drastically impact other consumers.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/RazorParser.cs]: `parse single directive method to RazorSyntaxTree correct` - this is exactly what I'm suggesting.

Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/RazorParser.cs]: Had a discussion offline and @rynowak came up with an awesomer reason to keep this on options. You can see options on the result of a parse, therefore you can identify an "only first directive" parse vs. a normal parse.

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/HtmlLanguageCharacteristics.cs [hunk 2]
```diff
@@ -8,17 +8,33 @@ namespace Microsoft.AspNetCore.Razor.Language.Legacy
 {
     internal class HtmlLanguageCharacteristics : LanguageCharacteristics<HtmlTokenizer, HtmlSymbol, HtmlSymbolType>
     {
+        private bool _useDirectiveTokenizer;
         private static readonly HtmlLanguageCharacteristics _instance = new HtmlLanguageCharacteristics();
+        private static readonly HtmlLanguageCharacteristics _instanceWithDirectiveTokenizer =
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/HtmlLanguageCharacteristics.cs]: Are you able to just sublcass `HtmlLanguageCharacteristics` and then ovveride CreateTokenizer?

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs [hunk 3]
```diff
@@ -88,7 +87,7 @@ public CSharpCodeParser(ParserContext context)
         }
 
         public CSharpCodeParser(IEnumerable<DirectiveDescriptor> directiveDescriptors, ParserContext context)
-            : base(CSharpLanguageCharacteristics.Instance, context)
+            : base(context.StopParsingAfterFirstDirective ? CSharpLanguageCharacteristics.InstanceWithDirectiveTokenizer : CSharpLanguageCharacteristics.Instance, context)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs]: `context.StopParsingAfterFirstDirective ? AjaysCoolSingleDirectiveCSharpLanguageCharacteristics.Instance : CSharpLanguageCharacteristics.Instance`

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/HtmlMarkupParser.cs [hunk 4]
```diff
@@ -38,7 +38,7 @@ internal class HtmlMarkupParser : TokenizerBackedParser<HtmlTokenizer, HtmlSymbo
         };
 
         public HtmlMarkupParser(ParserContext context)
-            : base(HtmlLanguageCharacteristics.Instance, context)
+            : base(context.StopParsingAfterFirstDirective ? HtmlLanguageCharacteristics.InstanceWithDirectiveTokenizer : HtmlLanguageCharacteristics.Instance, context)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/HtmlMarkupParser.cs]: `context.StopParsingAfterFirstDirective ? AjaysCoolSingleDirectiveHtmlLanguageCharacteristics.InstanceWithDirectiveTokenizer : HtmlLanguageCharacteristics.Instance`

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/RazorParser.cs [hunk 5]
```diff
@@ -37,7 +37,10 @@ public virtual RazorSyntaxTree Parse(RazorSourceDocument source)
 
             var reader = new SeekableTextReader(chars, source.FileName);
 
-            var context = new ParserContext(reader, Options.DesignTimeMode);
+            var context = new ParserContext(reader, Options.DesignTimeMode)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/RazorParser.cs]: Might as well just pass in the options now? If you don't see value, no need.

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/TokenizerBackedParser.cs [hunk 6]
```diff
@@ -25,6 +25,14 @@ protected TokenizerBackedParser(LanguageCharacteristics<TTokenizer, TSymbol, TSy
             Span = new SpanBuilder(CurrentLocation);
         }
 
+        protected TokenizerBackedParser(
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/TokenizerBackedParser.cs]: Do you need this guy?

**Comment by ajaybhargavb** [comment]
🆙 📅 No longer in design.

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/FirstDirectiveCSharpLanguageCharacteristics.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        private static readonly FirstDirectiveCSharpLanguageCharacteristics _instance = new FirstDirectiveCSharpLanguageCharacteristics();
+
+        private FirstDirectiveCSharpLanguageCharacteristics()
+        {
+        }
+
+        public new static FirstDirectiveCSharpLanguageCharacteristics Instance => _instance;
+
+        public override CSharpTokenizer CreateTokenizer(ITextDocument source)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/FirstDirectiveCSharpLanguageCharacteristics.cs]: Shame no pranav points

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/FirstDirectiveHtmlLanguageCharacteristics.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        private static readonly FirstDirectiveHtmlLanguageCharacteristics _instance = new FirstDirectiveHtmlLanguageCharacteristics();
+
+        private FirstDirectiveHtmlLanguageCharacteristics()
+        {
+        }
+
+        public new static FirstDirectiveHtmlLanguageCharacteristics Instance => _instance;
+
+        public override HtmlTokenizer CreateTokenizer(ITextDocument source)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/FirstDirectiveHtmlLanguageCharacteristics.cs]: Shame!

**Pull request merged**